src/sinks/clickhouse/arrow_schema.rs

```diff
@@ -0,0 +1,767 @@
+//! Schema fetching and Arrow type mapping for ClickHouse tables.
+
+use std::sync::Arc;
+
+use ::arrow::datatypes::{DataType, Field, Schema, TimeUnit};
+use async_trait::async_trait;
+use http::{Request, StatusCode};
+use hyper::Body;
+use serde::Deserialize;
+use vector_lib::codecs::encoding::format::{ArrowEncodingError, SchemaProvider};
+
+use crate::http::{Auth, HttpClient};
+
+const DECIMAL32_PRECISION: u8 = 9;
+const DECIMAL64_PRECISION: u8 = 18;
+const DECIMAL128_PRECISION: u8 = 38;
+const DECIMAL256_PRECISION: u8 = 76;
+
+#[derive(Debug, Deserialize)]
+struct ColumnInfo {
+    name: String,
+    #[serde(rename = "type")]
+    column_type: String,
+}
+
+/// Fetches the schema for a ClickHouse table and converts it to an Arrow schema.
+pub async fn fetch_table_schema(
+    client: &HttpClient,
+    endpoint: &str,
+    database: &str,
+    table: &str,
+    auth: Option<&Auth>,
+) -> crate::Result<Arc<Schema>> {
+    // Query to get table schema
+    let query = format!(
+        "SELECT name, type \
+         FROM system.columns WHERE database = '{}' AND table = '{}' \
+         ORDER BY position FORMAT JSONEachRow",
+        database, table
+    );
```

Comment:
For an abundance of safety here I think we should have some kind of SQLi prevention. Something as simple as limiting database and table to ascii chars and `-, _` should be enough. I wouldn't be opposed to using some other crate here instead.

Note: I spent some time looking if the `sqlx` crate is viable for this but it doesn't look like it is unfortunately. All query operations depend on a database connection.